crates/ruma-common/src/serde.rs

```
@@ -70,17 +70,26 @@ pub fn is_true(b: &bool) -> bool {
 }
 
 /// Returns None if the serialization fails
-pub fn or_empty<'de, D: Deserializer<'de>, T: Deserialize<'de>>(
+pub fn or_empty<'de, D: Deserializer<'de>, T: for <'a> Deserialize<'a>>(
         deserializer: D,
     ) -> Result<Option<T>, D::Error> {
-        #[derive(serde::Deserialize)]
-        #[serde(untagged)]
-        enum OrEmpty<T> {
-            NotEmpty(T),
-            Empty {},
+        let json = Box::<RawJsonValue>::deserialize(deserializer)?;
+
+        let res = serde_json::from_str::<Option<T>>(json.get()).map_err(de::Error::custom);
+
+        match res {
+            Ok(a) => Ok(a),
+            Err(e) => {
+                #[derive(Deserialize)]
+                #[serde(deny_unknown_fields)]
```

This solution seems to work better, but I'm a little unsure about this part, since it'll fail if there's a new field 🧐

---

this is to ensure that it only treats an *empty* object as null. otherwise it will fail silently on any object that fails to parse into the request type and treat it as null

---

So just to confirm, if Element one day decides to send a random junk/unknown field it won't break deserialisation on anything here?

---

it will not break on extra fields if it is still valid AuthData (or T, if this is used anywhere else) otherwise. if element sends an object with *only* random junk, then yeah, this will error, but it probably should in that case. this pr is designed to treat only *empty* objects as null, not treat *all* incorrect data as null